cpp/src/arrow/dataset/discovery.cc

```diff
@@ -156,27 +177,26 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
 
   ARROW_ASSIGN_OR_RAISE(auto files, filesystem->GetFileInfo(selector));
 
-  std::vector<std::string> paths;
+  // XXX We may avoid copying around vectors of FileInfo by filtering in place.
+  std::vector<fs::FileInfo> filtered_files;
```

Comment:
```suggestion
  auto files_end = std::remove_if(files.begin(), files.end(), [&](const FileInfo& info) {
    if (!info.IsFile()) {
      return true;
    }

    if (StartsWithAnyOf(info.path(), options.selector_ignore_prefixes)) {
      return true;
    }
    
    return false;
  });
  files.erase(files_end, files.end());
```

Reply:
Yeah, that's what I thought :-) Thanks for spelling it out, will give it a try.